File: Emoticon/UITextView+Extension.m
//
//  UITextView+Extension.m
//  表情键盘OC
//
//  Created by 单车 on 2019/10/30.
//  Copyright © 2019 单车. All rights reserved.
//

#import "UITextView+Extension.h"
#import "EmoticonAttachement.h"
#import "Emoticon.h"
@implementation UITextView (Extension)
- (NSString *)getEmoticonString{
    //1.获取属性字符串
    NSMutableAttributedString *attrMStr = [[NSMutableAttributedString alloc]initWithAttributedString:self.attributedText];
    
    //2.遍历属性字符串
    NSRange range = NSMakeRange(0, attrMStr.length);
    
    //遍历获得符合指定属性或属性字典的区域(range)，并在 block 中进行设置
    [attrMStr enumerateAttributesInRange:range options:NSAttributedStringEnumerationLongestEffectiveRangeNotRequired usingBlock:^(NSDictionary<NSAttributedStringKey,id> * _Nonnull attrs, NSRange range, BOOL * _Nonnull stop) {
        EmoticonAttachement *attachement = attrs[@"NSAttachment"];
        if (attachement != nil) {
            [attrMStr replaceCharactersInRange:range withString:attachement.chs];
        }
    }];
    return attrMStr.string;
}

- (void)insertEmoticon:(Emoticon *)emoticon{
    if (emoticon.isEmpty == YES) {
        return;
    }
    if (emoticon.isRemove) {
        [self deleteBackward];
        return;
    }
    
    if (emoticon.emojiCode != nil) {
        UITextRange *textRange = self.selectedTextRange;
        
        [self replaceRange:textRange withText:emoticon.emojiCode];
        
        return;
    }
    //4.普通表情：图文混排
    //4.1根据图片路径创建属性字符串
    EmoticonAttachement *attachement = [[EmoticonAttachement alloc]init];
    attachement.chs = emoticon.chs;
    attachement.image = [UIImage imageWithContentsOfFile:emoticon.pngPath];
    UIFont *font = self.font;
    attachement.bounds = CGRectMake(0, -4, font.lineHeight, font.lineHeight);
    NSAttributedString *attrImageStr = [NSAttributedString attributedStringWithAttachment:attachement];
    
    //4.2 创建可变的属性字符串
    NSMutableAttributedString *attrMStr = [[NSMutableAttributedString alloc]initWithAttributedString:self.attributedText];
    
    NSRange range = [self selectedRangeInTextView:self];
        
    [attrMStr replaceCharactersInRange:range withAttributedString:attrImageStr];
    
    self.attributedText = attrMStr;
    
    self.font = font;
    
    self.selectedRange = NSMakeRange(range.location + 1, 0);
}

- (NSRange) selectedRangeInTextView:(UITextView*)textView
{
    UITextPosition* beginning = textView.beginningOfDocument;

    UITextRange* selectedRange = textView.selectedTextRange;
    UITextPosition* selectionStart = selectedRange.start;
    UITextPosition* selectionEnd = selectedRange.end;

    const NSInteger location = [textView offsetFromPosition:beginning toPosition:selectionStart];
    const NSInteger length = [textView offsetFromPosition:selectionStart toPosition:selectionEnd];

    return NSMakeRange(location, length);
}
@end

File: 09-表情键盘备份/09-表情键盘/Emoticon/UITextView+Extension.swift
//
//  UITextView+Extension.swift
//  09-表情键盘
//
//  Created by 单车 on 2019/10/23.
//  Copyright © 2019 单车. All rights reserved.
//

import UIKit

extension UITextView{
//    func insertEmoticon(emoticon:Emoticon){
//        //1.空白表情
//        if emoticon.isEmpty {
//            return
//        }
//        
//        //2.删除按钮
//        if emoticon.isRemove {
//            textView.deleteBackward()
//            return
//        }
//        
//        //3.emoji表情
//        if emoticon.emojiCode != nil {
//            //3.1获取光标的位置 uitextRange
//            let textRange = textView.selectedTextRange!
//            
//            //3.2 替换emoji 表情
//            textView.replace(textRange, withText: emoticon.emojiCode!)
//            
//            return
//        }
//        //4.普通表情：图文混排
//        //4.1根据图片路径创建属性字符串
//        let attachment = EmoticonAttachment()
//        attachment.chs = emoticon.chs
//        attachment.image = UIImage(contentsOfFile: emoticon.pngPath!)
//        let font = textView.font!
//        
//        attachment.bounds = CGRect(x: 0, y: -4, width: font.lineHeight, height: font.lineHeight)
//        let attrImageStr = NSAttributedString(attachment: attachment)
//        
//        //4.2创建可变的属性字符串
//        let attrMStr = NSMutableAttributedString(attributedString: textView.attributedText)
//        
//        //获取光标所在的位置
//        let range = textView.selectedRange
//        
//        //4.3将图片属性字符串替换到可变属性字符串的某一个位置
//        attrMStr.replaceCharacters(in: range, with: attrImageStr)
//        
//        //显示属性字符串
//        textView.attributedText = attrMStr
//        
//        //将文字的大小重置
//        textView.font = font
//        
//        //将光标设置回原来的位置+1
//        textView.selectedRange = NSRange(location: range.location + 1, length: 0)
//    }
}
